csr_check.go

```diff
@@ -162,3 +180,77 @@ func authorizeCSR(machineList *v1beta1.MachineList, req *certificatesv1beta1.Cer
 
 	return nil
 }
+
+func authorizeNodeClientCSR(machines []v1beta1.Machine, nodes corev1client.NodeInterface, req *certificatesv1beta1.CertificateSigningRequest, csr *x509.CertificateRequest) error {
+	if !isReqFromNodeBootstrapper(req) {
+		return fmt.Errorf("CSR %s for node client cert has wrong user", req.Name)
+	}
+
+	nodeName := strings.TrimPrefix(csr.Subject.CommonName, nodeUserPrefix)
+	if len(nodeName) == 0 {
+		return fmt.Errorf("CSR %s has empty node name", req.Name)
+	}
+
+	_, err := nodes.Get(nodeName, metav1.GetOptions{})
+	switch {
+	case err == nil:
+		return fmt.Errorf("node %s already exists", nodeName)
+	case errors.IsNotFound(err):
+		// good, node does not exist
+	default:
+		return fmt.Errorf("failed to check if node %s already exists: %v", nodeName, err)
+	}
+
+	nodeMachine, ok := findMatchingMachineFromInternalDNS(nodeName, machines)
+	if !ok {
+		return fmt.Errorf("failed to find machine for node %s", nodeName)
+	}
+
+	if nodeMachine.Status.NodeRef != nil {
+		return fmt.Errorf("machine for node %s already has node ref", nodeName)
+	}
+
+	if !isWorkerMachine(nodeMachine) {
+		return fmt.Errorf("machine for node %s is not a worker", nodeName)
+	}
+
+	start := nodeMachine.CreationTimestamp.Add(-maxMachineDelta) // TODO maybe this should be really small to account only for clock drift
+	end := nodeMachine.CreationTimestamp.Add(maxMachineDelta)
+	if !inTimeSpan(start, end, req.CreationTimestamp.Time) {
+		return fmt.Errorf("CSR %s creation time %s not in range (%s, %s)", req.Name, req.CreationTimestamp.Time, start, end)
+	}
+
+	return nil // approve node client cert
+}
+
+func isReqFromNodeBootstrapper(req *certificatesv1beta1.CertificateSigningRequest) bool {
+	return req.Spec.Username == nodeBootstrapperUsername && nodeBootstrapperGroups.Equal(sets.NewString(req.Spec.Groups...))
+}
+
+func findMatchingMachineFromNodeRef(nodeName string, machines []v1beta1.Machine) (v1beta1.Machine, bool) {
+	for _, machine := range machines {
+		if machine.Status.NodeRef != nil && machine.Status.NodeRef.Name == nodeName {
+			return machine, true
+		}
+	}
+	return v1beta1.Machine{}, false
+}
+
+func findMatchingMachineFromInternalDNS(nodeName string, machines []v1beta1.Machine) (v1beta1.Machine, bool) {
+	for _, machine := range machines {
```

Comment:
I'm not sure this is something we can guarantee being available for every cloud. Not sure `NodeInternalDNS` is exposed on the API in Azure/GCP. It might be only exposed via metadata.

Reply:
Any other way to very specifically find the machine?

Reply:
would the privateIP be sufficient?
medium term we should include the node/machine providerID here but this is not supported today

Reply:
The CSR only includes the (future) node name so we have to base the check on that...

Reply:
couldn't we get the privateIP here (in future providerID as well) https://github.com/openshift/cluster-machine-approver/pull/26/files#diff-a482a1a664767249971b0b71ca9baaf0R194 and pass it down to `findMatchingMachineBy*()`?